doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+
+The block types are executed in the following order:
+
+* [`precalculate`](#precalculatevariable_name-block)
+* [`on_response`](#on_responseblock)
+* [`next_node_calculation`](#next_node_calculationvariable_name-block)
+* [`validate`](#validatemessage_key-block)
+* [`next_node`](#next_nodeblock)
+* [`save_input_as`](#save_input_asvariable_name)
+* [`calculate`](#calculatevariable_name-block)
+
+Each of these block types and the point at which they are executed is explained in more detail below:
+
+##### `precalculate(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed to render a question or outcome node template.
+* It's simplest to think about these blocks executing before the user response is parsed from the request path.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> Note that due to an oversight in the implementation of `Flow#process`, it's not currently possible to use a `precalculate` block in the first question.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+##### `on_response(&block)`
+
+* These blocks are intended to be used to store user responses on a `calculator` state variable. They are a [relatively new addition to the DSL][introduction-of-on-response-blocks].
+* These blocks are called  after the user response has been parsed from the request path and before any of the `next_node_calculation` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is not used and no state variable is stored.
+
+> The use of these blocks is encouraged; however, they should only ever be used to store a single, `calculator` state variable (in the first question definition); otherwise they should only be used to store user responses on that `calculator` object.
+
+##### `next_node_calculation(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed in the `next_node` block.
+* These blocks are executed after all the `on_response` blocks have been executed and before any of the `validate` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
```

Comment:
> This should be included in the docs somewhere. I don't recall reading it anywhere, and I wasn't aware that this is where the code is heading.

I think this would be better addressed by improving the [refactoring documentation][1]. I want to keep this PR more focussed on the current behaviour and the [flow-definition documentation][2]. And as I've said elsewhere, I think this is something that would benefit from discussing with the whole team ideally in-front of a white-board.

[1]: https://github.com/alphagov/smart-answers/blob/improve-flow-definition-documentation/doc/refactoring.md
[2]: https://github.com/alphagov/smart-answers/blob/improve-flow-definition-documentation/doc/flow-definition.md